core/services/workflows/monitoring.go

```diff
@@ -0,0 +1,227 @@
+	}
+
+	// Add new key-value pairs
+	for i := 0; i < len(keyValues); i += 2 {
+		key := keyValues[i]
+		value := keyValues[i+1]
+		newCustomMessageLabeler.labels[key] = value
+	}
+
+	return newCustomMessageLabeler
+}
+
+// sendLogAsCustomMessage emits a KeystoneCustomMessage with msg and labels as data.
+// any key in labels that is not part of orderedLabelKeys will not be transmitted
+func (c customMessageLabeler) sendLogAsCustomMessage(msg string) error {
+	return sendLogAsCustomMessageW(msg, c.labels)
+}
+
+type customMetricsLabeler struct {
+	labels map[string]string
+}
+
+func NewCustomMetricsLabeler() customMetricsLabeler {
+	return customMetricsLabeler{labels: make(map[string]string)}
+}
+
+// with adds multiple key-value pairs to the customMessageLabeler for transmission with sendLogAsCustomMessage
+func (c customMetricsLabeler) with(keyValues ...string) customMetricsLabeler {
+	newCustomMetricsLabeler := NewCustomMetricsLabeler()
+
+	if len(keyValues)%2 != 0 {
+		// If an odd number of key-value arguments is passed, return the original customMessageLabeler unchanged
+		return c
+	}
+
+	// Copy existing labels from the current agent
+	for k, v := range c.labels {
+		newCustomMetricsLabeler.labels[k] = v
+	}
+
+	// Add new key-value pairs
+	for i := 0; i < len(keyValues); i += 2 {
+		key := keyValues[i]
+		value := keyValues[i+1]
+		newCustomMetricsLabeler.labels[key] = value
+	}
+
+	return newCustomMetricsLabeler
+}
+
+func (c customMetricsLabeler) incrementRegisterTriggerFailureCounter(ctx context.Context) {
+	otelLabels := kvMapToOtelAttributes(c.labels)
+	registerTriggerFailureCounter.Add(ctx, 1, metric.WithAttributes(otelLabels...))
+}
+
+func (c customMetricsLabeler) incrementCapabilityInvocationCounter(ctx context.Context) {
+	otelLabels := kvMapToOtelAttributes(c.labels)
+	capabilityInvocationCounter.Add(ctx, 1, metric.WithAttributes(otelLabels...))
+}
+
+func (c customMetricsLabeler) updateWorkflowExecutionLatencyGauge(ctx context.Context, val int64) {
+	otelLabels := kvMapToOtelAttributes(c.labels)
+	workflowExecutionLatencyGauge.Record(ctx, val, metric.WithAttributes(otelLabels...))
+}
+
+func (c customMetricsLabeler) incrementTotalWorkflowStepErrorsCounter(ctx context.Context) {
+	otelLabels := kvMapToOtelAttributes(c.labels)
+	workflowStepErrorCounter.Add(ctx, 1, metric.WithAttributes(otelLabels...))
+}
+
+func (c customMetricsLabeler) updateTotalWorkflowsGauge(ctx context.Context, val int64) {
+	otelLabels := kvMapToOtelAttributes(c.labels)
+	workflowsRunningGauge.Record(ctx, val, metric.WithAttributes(otelLabels...))
+}
+
+// Observability keys
+const (
+	cIDKey  = "capabilityID"
+	tIDKey  = "triggerID"
+	wIDKey  = "workflowID"
+	eIDKey  = "workflowExecutionID"
+	wnKey   = "workflowName"
+	woIDKey = "workflowOwner"
+	sIDKey  = "stepID"
+	sRKey   = "stepRef"
+)
+
+var orderedLabelKeys = []string{sRKey, sIDKey, tIDKey, cIDKey, eIDKey, wIDKey}
+
+var labelsMap = make(map[string]interface{})
```

Comment:
/nit current idiom is `any` rather than `interface{}`

Reply:
Thank you